In [ ]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler,OneHotEncoder
import pandas as pd
import tensorflow as tf
from google.colab import drive
drive.mount('/content/gdrive')

Drive already mounted at /content/gdrive; to attempt to forcibly remount, call drive.mount("/content/gdrive", force_remount=True).


In [ ]:
path = "gdrive/My Drive/Colab Notebooks/Deep_Learning_Week/data/HR-Employee-Attrition.csv"
hr_df = pd.read_csv(path)
hr_df.head()

Age Attrition     BusinessTravel  DailyRate              Department  \
0   41       Yes      Travel_Rarely       1102                   Sales   
1   49        No  Travel_Frequently        279  Research & Development   
2   37       Yes      Travel_Rarely       1373  Research & Development   
3   33        No  Travel_Frequently       1392  Research & Development   
4   27        No      Travel_Rarely        591  Research & Development   

   DistanceFromHome  Education EducationField  EmployeeCount  EmployeeNumber  \
0                 1          2  Life Sciences              1               1   
1                 8          1  Life Sciences              1               2   
2                 2          2          Other              1               4   
3                 3          4  Life Sciences              1               5   
4                 2          1        Medical              1               7   

   ...  RelationshipSatisfaction StandardHours  StockOptionLevel  \
0  ...                         1            80                 0   
1  ...                         4            80                 1   
2  ...                         2            80                 0   
3  ...                         3            80                 0   
4  ...                         4            80                 1   

   TotalWorkingYears  TrainingTimesLastYear WorkLifeBalance  YearsAtCompany  \
0                  8                      0               1               6   
1                 10                      3               3              10   
2                  7                      3               3               0   
3                  8                      3               3               8   
4                  6                      3               3               2   

  YearsInCurrentRole  YearsSinceLastPromotion  YearsWithCurrManager  
0                  4                        0                     5  
1                  7                        1                     7  
2                  0                        0                     0  
3                  7                        3                     0  
4                  2                        2                     2  

[5 rows x 35 columns]

In [ ]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 35 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EmployeeNumber            1470 non-null   int64 
 10  EnvironmentSatisfaction   1470 non-null   int64 
 11  Gender                    1470 non-null   object
 12  HourlyRate                1470 non-null   int64 
 13  JobInvolvement            1470 non-null   int64 
 14  JobLevel                

In [ ]:
# grab all objects and put them in a list
attrition_cat = hr_df.dtypes[hr_df.dtypes == 'object'].index.to_list()

In [ ]:
hr_df[attrition_cat].nunique()

Attrition         2
BusinessTravel    3
Department        3
EducationField    6
Gender            2
JobRole           9
MaritalStatus     3
Over18            1
OverTime          2
dtype: int64

In [ ]:
# OneHotEncode cat data
enc = OneHotEncoder(sparse=False)

# Fit & transform the enc using the cat list
encode_df = pd.DataFrame(enc.fit_transform(hr_df[attrition_cat]))

# Add encoded vars to DF
encode_df.columns = enc.get_feature_names_out(attrition_cat)
encode_df.head()

Attrition_No  Attrition_Yes  BusinessTravel_Non-Travel  \
0           0.0            1.0                        0.0   
1           1.0            0.0                        0.0   
2           0.0            1.0                        0.0   
3           1.0            0.0                        0.0   
4           1.0            0.0                        0.0   

   BusinessTravel_Travel_Frequently  BusinessTravel_Travel_Rarely  \
0                               0.0                           1.0   
1                               1.0                           0.0   
2                               0.0                           1.0   
3                               1.0                           0.0   
4                               0.0                           1.0   

   Department_Human Resources  Department_Research & Development  \
0                         0.0                                0.0   
1                         0.0                                1.0   
2                         0.0                                1.0   
3                         0.0                                1.0   
4                         0.0                                1.0   

   Department_Sales  EducationField_Human Resources  \
0               1.0                             0.0   
1               0.0                             0.0   
2               0.0                             0.0   
3               0.0                             0.0   
4               0.0                             0.0   

   EducationField_Life Sciences  ...  JobRole_Research Director  \
0                           1.0  ...                        0.0   
1                           1.0  ...                        0.0   
2                           0.0  ...                        0.0   
3                           1.0  ...                        0.0   
4                           0.0  ...                        0.0   

   JobRole_Research Scientist  JobRole_Sales Executive  \
0                         0.0                      1.0   
1                         1.0                      0.0   
2                         0.0                      0.0   
3                         1.0                      0.0   
4                         0.0                      0.0   

   JobRole_Sales Representative  MaritalStatus_Divorced  \
0                           0.0                     0.0   
1                           0.0                     0.0   
2                           0.0                     0.0   
3                           0.0                     0.0   
4                           0.0                     0.0   

   MaritalStatus_Married  MaritalStatus_Single  Over18_Y  OverTime_No  \
0                    0.0                   1.0       1.0          0.0   
1                    1.0                   0.0       1.0          1.0   
2                    0.0                   1.0       1.0          0.0   
3                    1.0                   0.0       1.0          0.0   
4                    1.0                   0.0       1.0          1.0   

   OverTime_Yes  
0           1.0  
1           0.0  
2           1.0  
3           1.0  
4           0.0  

[5 rows x 31 columns]

In [ ]:
# Merge the encoded dataframe with the OG dataframe, then drop the categorical columns
hr_df = hr_df.merge(encode_df, left_index=True, right_index=True)
hr_df = hr_df.drop(attrition_cat, 1)
hr_df.head()

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


Age  DailyRate  DistanceFromHome  Education  EmployeeCount  EmployeeNumber  \
0   41       1102                 1          2              1               1   
1   49        279                 8          1              1               2   
2   37       1373                 2          2              1               4   
3   33       1392                 3          4              1               5   
4   27        591                 2          1              1               7   

   EnvironmentSatisfaction  HourlyRate  JobInvolvement  JobLevel  ...  \
0                        2          94               3         2  ...   
1                        3          61               2         2  ...   
2                        4          92               2         1  ...   
3                        4          56               3         1  ...   
4                        1          40               3         1  ...   

   JobRole_Research Director  JobRole_Research Scientist  \
0                        0.0                         0.0   
1                        0.0                         1.0   
2                        0.0                         0.0   
3                        0.0                         1.0   
4                        0.0                         0.0   

   JobRole_Sales Executive  JobRole_Sales Representative  \
0                      1.0                           0.0   
1                      0.0                           0.0   
2                      0.0                           0.0   
3                      0.0                           0.0   
4                      0.0                           0.0   

   MaritalStatus_Divorced  MaritalStatus_Married  MaritalStatus_Single  \
0                     0.0                    0.0                   1.0   
1                     0.0                    1.0                   0.0   
2                     0.0                    0.0                   1.0   
3                     0.0                    1.0                   0.0   
4                     0.0                    1.0                   0.0   

   Over18_Y  OverTime_No  OverTime_Yes  
0       1.0          0.0           1.0  
1       1.0          1.0           0.0  
2       1.0          0.0           1.0  
3       1.0          0.0           1.0  
4       1.0          1.0           0.0  

[5 rows x 57 columns]

In [ ]:
hr_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1470 entries, 0 to 1469
Data columns (total 57 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Age                                1470 non-null   int64  
 1   DailyRate                          1470 non-null   int64  
 2   DistanceFromHome                   1470 non-null   int64  
 3   Education                          1470 non-null   int64  
 4   EmployeeCount                      1470 non-null   int64  
 5   EmployeeNumber                     1470 non-null   int64  
 6   EnvironmentSatisfaction            1470 non-null   int64  
 7   HourlyRate                         1470 non-null   int64  
 8   JobInvolvement                     1470 non-null   int64  
 9   JobLevel                           1470 non-null   int64  
 10  JobSatisfaction                    1470 non-null   int64  
 11  MonthlyIncome                      1470 non-null   int64

In [ ]:
# Split the data into input (X) and output/target (y)
y = hr_df['Attrition_Yes'].values
X = hr_df.drop(['Attrition_Yes','Attrition_No'],1).values

# split the preprocessed data into testing and training sets
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=78)


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:3: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  This is separate from the ipykernel package so we can avoid doing imports until


In [ ]:
# Scale the data with StandardScaler
scaler = StandardScaler()

# fit the scaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)


## Build the NN Model

In [ ]:
# Define the model - DEEP NEURAL NET YO
input_features = len(X_train[0]) # must equal number of variables in feature df
hidden_nodes_1 = 8
hidden_nodes_2 = 5

nn = tf.keras.models.Sequential()

# first hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_1, input_dim=input_features, activation='relu'))

# 2nd hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_2, activation='relu'))

# output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check model structure
nn.summary()


Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 8)                 448       
                                                                 
 dense_1 (Dense)             (None, 5)                 45        
                                                                 
 dense_2 (Dense)             (None, 1)                 6         
                                                                 
Total params: 499
Trainable params: 499
Non-trainable params: 0
_________________________________________________________________


#### Number of weight parameters = number of input values (55 in this case) times number of neurons. 

First layer is: 55 x 8 + 8 for bias = 448

Second layer is: 8 (from first layer) x 5 neurons for 2nd layer = 45

In [ ]:
# compile the model and define loss & accuracy metrics
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [ ]:
# Train the model
fit_model = nn.fit(X_train, y_train, epochs=100)

Epoch 1/100
35/35 [==============================] - 2s 4ms/step - loss: 733.1397 - accuracy: 0.8276
Epoch 2/100
35/35 [==============================] - 0s 3ms/step - loss: 384.2342 - accuracy: 0.8258
Epoch 3/100
35/35 [==============================] - 0s 2ms/step - loss: 142.3090 - accuracy: 0.7877
Epoch 4/100
35/35 [==============================] - 0s 3ms/step - loss: 26.5750 - accuracy: 0.7377
Epoch 5/100
35/35 [==============================] - 0s 3ms/step - loss: 5.2575 - accuracy: 0.7260
Epoch 6/100
35/35 [==============================] - 0s 3ms/step - loss: 1.4133 - accuracy: 0.6425
Epoch 7/100
35/35 [==============================] - 0s 3ms/step - loss: 0.8349 - accuracy: 0.7015
Epoch 8/100
35/35 [==============================] - 0s 3ms/step - loss: 0.6351 - accuracy: 0.7142
Epoch 9/100
35/35 [==============================] - 0s 4ms/step - loss: 0.5666 - accuracy: 0.7278
Epoch 10/100
35/35 [==============================] - 0s 3ms/step - loss: 0.5605 - accuracy: 0.7368
Ep

In [ ]:
# Evaluate model performance
model_loss, model_accuracy = nn.evaluate(X_test, y_test, verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

12/12 - 0s - loss: 0.3935 - accuracy: 0.8723 - 340ms/epoch - 28ms/step
Loss: 0.3935394287109375, Accuracy: 0.8722826242446899
